### PR TITLE
Remove the tie between event counters and logging

### PR DESCRIPTION
**Pull request by tristan957**
When something was being logged, it would usually also be accompanied by
an event counter nearby, so this led to two event counters essentially
counting the same things. It is better suited for event counters to
exist separately from logging in order to avoid this hidden event
counter duplication.

Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

**Comment by tristan957** [comment]
This is a side effect of this PR. Note this was a debugoptimized build.

```
$ ./build/cli/hse storage info /media/hse-tests/ 2> /dev/null
[HSE] lib/config/argv.c:101 argv_deserialize_to_params: Applying HSE global param logging.enabled from paramv
[HSE] lib/util/src/platform.c:215 hse_platform_init: version r3.0.0-dev-77-ga5d92e395-dirty, program ./build/cli/hse
[HSE] lib/util/src/platform.c:198 hse_cpu_init: bogomips 4799, freq 2399500000, shift 21, mult 873995, L1D_CLSZ 64
[HSE] lib/binding/kvdb_interface.c:142 hse_init: Memory available: 244 GiB
[HSE] lib/kvs/kvs_cursor.c:1431 kvs_curcache_init: bktsz 61440, bktc 28, nperbkt 892
[HSE] lib/binding/kvdb_interface.c:154 hse_init: Rest server started: /tmp/hse-2106982.sock
[HSE] lib/config/config.c:540 config_from_kvdb_conf: No config file (/media/hse-tests//kvdb.conf)
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 1 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 2 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 3 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 4 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 5 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 6 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 7 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 8 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 9 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 10 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 11 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 12 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 13 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 14 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 15 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 16 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 17 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 18 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 19 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 20 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 21 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 22 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 23 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 24 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 25 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 26 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 27 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 28 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 29 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 30 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 31 found 0 valid mblocks, uniq 0
[HSE] lib/mpool/src/mblock_file.c:522 mblock_file_meta_load: mclass 1, file-id 32 found 0 valid mblocks, uniq 0
[HSE] lib/kvdb/ikvdb.c:1508 ikvdb_open: setting throttling.init_policy to "heavy" for KVDB(/media/hse-tests/)
[HSE] lib/kvdb/throttle.c:103 throttle_init_params: delay 2097151 u_ms 25 rcycles 200 icycles 8 scycles 20 dcycles 32
[HSE] lib/cn/csched_sp3.c:468 sp3_refresh_samp: sp3 kvdb_rparam csched_samp_max changed from 0 to 150
[HSE] lib/cn/csched_sp3.c:477 sp3_refresh_samp: sp3 kvdb_rparam csched_lo_th_pct changed from 0 to 25
[HSE] lib/cn/csched_sp3.c:486 sp3_refresh_samp: sp3 kvdb_rparam csched_hi_th_pct changed from 0 to 75
[HSE] lib/cn/csched_sp3.c:495 sp3_refresh_samp: sp3 kvdb_rparam csched_leaf_pct changed from 0 to 90
[HSE] lib/cn/csched_sp3.c:501 sp3_refresh_samp: sp3 new samp input params: samp 150, lwm_pct 25, hwm_pct 75, leaf_pct 90
[HSE] lib/cn/csched_sp3.c:552 sp3_refresh_samp: sp3 samp derived params: samp lo/hi/max: 1.179 1.375 1.500 good/leaf ratio min/lo/hi: 0.733 0.933 0.800
[HSE] lib/cn/csched_sp3.c:651 sp3_refresh_thresholds: sp3 thresholds: rspill: min/max/wlenmb 5/9/8192, lcomp: max/pct/keys 12/100%/256, llen: min/max 4/8, idlec: 2, idlem: 10, lscat: hwm/max 64/3 split 5
[HSE] lib/c0/c0sk.c:524 c0sk_open: c0sk_open(0) complete
[HSE] lib/wal/wal.c:729 wal_open: setting durability.mclass policy to "capacity"
MEDIA_CLASS     ALLOCATED_BYTES USED_BYTES      PATH                     
capacity        67248128        67248128        /media/hse-tests/capacity
[HSE] lib/c0/c0sk_internal.c:490 c0sk_merge_loop: (140216916484096) Entries added: cn 0 lc 0
[HSE] lib/c0/c0sk_internal.c:490 c0sk_merge_loop: (140216916484096) Entries added: cn 0 lc 0
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (0 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (1 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (2 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (3 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (4 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (5 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (6 - 4096, 4096 : 4096 : 4096)
[HSE] lib/wal/wal_buffer.c:337 wal_bufset_stats_dump: WAL closing - Offsets (7 - 4096, 4096 : 4096 : 4096)
```
Not really sure what we should do about this.

**Comment by tristan957** [comment]
Also I found a deadlock. If `vlb_fini()` is called before `vlb_init()`, which occurs in the event `hse_platform_init()` errors out prior to `vlb_init()` or `vlb_init()` errors out, then I assume we either enter undefined behavior or a deadlock locking a spinlock that has never been initialized.

**Comment by tristan957** [comment]
I haven't gotten rid of all the fprintf()s in libhse yet. On my list to do tomorrow in addition to solving the above problems.

**Comment by beckerg** [comment]
it's not a deadlock, hse_init() is broken:  It should only call hse_platform_fini() is it successfully called hse_platform_init().

File: lib/logging/include/hse/logging/logging.h [hunk 1]
```diff
@@ -19,6 +21,12 @@
 #define LOG_DEFAULT LOG_DEBUG
 #endif
 
+#ifdef HSE_REL_SRC_DIR
+#define LOG_FILE(x) (x + sizeof(HSE_REL_SRC_DIR) - 1)
+#else
+#define LOG_FILE(x) x
+#endif
```
Comment (beckerg) [inline lib/logging/include/hse/logging/logging.h:28]: Use a descriptive arg name with a single leading underscore.

Comment (beckerg) [reply lib/logging/include/hse/logging/logging.h:28]: It's not clear to me that calling this macro with anything other than __FILE__ makes sense...

File: lib/util/include/hse_util/vlb.h [hunk 2]
```diff
@@ -57,7 +57,7 @@ void *vlb_alloc(size_t sz);
  */
 void vlb_free(void *mem, size_t used);
 
-merr_t vlb_init(void) HSE_COLD;
+void vlb_init(void) HSE_COLD;
```
Comment (beckerg) [inline lib/util/include/hse_util/vlb.h:60]: I think it's a cleaner model if all the init functions are plumbed to return a merr, whether or not they need to...  Makes adding code to them down the road much easier....

File: lib/util/src/platform.c [hunk 3]
```diff
@@ -220,10 +216,7 @@ hse_platform_init(void)
 
     dt_init();
     event_counter_init();
-
-    err = logging_init(&hse_gparams.gp_logging);
-    if (err)
-        return err;
+    vlb_init();
```
Comment (beckerg) [inline lib/util/src/platform.c:219]: Don't move this, it prevents using anything initialized by hse_cpu_init() or in vlb_init().
Probably hse_cpu_init() should be called first of all these...

File: tests/unit/mpool/mblock_test.c [hunk 4]
```diff
@@ -347,7 +348,7 @@ mblock_rw(struct mpool *mp, uint64_t mbid, void *buf, size_t len, u64 boff, bool
     iovc = 0;
 
     /* Is the caller's buffer page aligned? */
-    if ((u64)pos & (u64)(PAGE_SIZE - 1)) {
+    if ((uintptr_t)pos & (uint64_t)(PAGE_SIZE - 1)) {
```
Comment (beckerg) [inline tests/unit/mpool/mblock_test.c:351]: Drive-by cleanup?

**Review by beckerg — APPROVED** [review]

**Comment by alexttx** [comment]
re: output using CLI, how about setting log level to warn and file=stderr when using cli?  verbose can set it to debug.

File: tests/unit/mpool/mblock_test.c (hunk 4 above)
Comment (tristan957) [reply tests/unit/mpool/mblock_test.c:351]: Nah, the includes are totally f-ed up, so when event counter no longer got included by logging, this was all the fallout. It took my liked an hour to slog through this. Totally annoying. This is why files should include what they use!

File: lib/logging/lib/logging.c [hunk 5]
```diff
@@ -49,12 +43,16 @@ logging_init(const struct logging_params *const params)
...
         return merr(EINVAL);
 
-    if (!params->lp_enabled)
-        return 0;
-
     if (params->lp_level < LOG_EMERG || params->lp_level > LOG_DEBUG)
         return merr(EINVAL);
 
+    if (params->lp_destination < LOG_DEST_STDOUT || params->lp_destination > LOG_DEST_SYSLOG)
+        return merr(EINVAL);
```
Comment (alexttx) [inline lib/logging/lib/logging.c:50]: Don't you have `LOG_DEST_MIN` and `_MAX`?  Alternatively, (1) get rid of `_MIN` and `_MAX` and use `0 <= v < LOG_DEST_COUNT`, or (2) use a switch.

File: lib/util/include/hse_util/vlb.h (hunk 2 above)
Comment (tristan957) [reply lib/util/include/hse_util/vlb.h:60]: I'll change it back, but you could just make this argument for function in my opinion. I don't see the difficulty in adding changing the return type 😄

File: lib/util/src/platform.c (hunk 3 above)
Comment (tristan957) [reply lib/util/src/platform.c:219]: I will move hse_cpu_init() up then.

File: lib/logging/lib/logging.c (hunk 5 above)
Comment (tristan957) [reply lib/logging/lib/logging.c:50]: oh yeah, I do. Whoops :).

File: lib/logging/include/hse/logging/logging.h (hunk 1 above)
Comment (tristan957) [reply lib/logging/include/hse/logging/logging.h:28]: The only thing I don't like here is that I have repeated the macro 3 times, logging, ev, perfc. Any thoughts on where a good play to put this macro is? `compiler.h`

Comment (beckerg) [reply lib/logging/include/hse/logging/logging.h:28]: Yeah, I saw that :-)   Yeah, it's definitely a buildenv/compiler thing, so that makes sense...

File: lib/util/include/hse_util/vlb.h (hunk 2 above)
Comment (beckerg) [reply lib/util/include/hse_util/vlb.h:60]: Ok for now, just an annoyance when the day comes I want to add something to the init() functions that return an error.  Used to be an extremely horrible ordeal because merr wasn't plumbed all the way down...

File: lib/util/src/platform.c (hunk 3 above)
Comment (beckerg) [reply lib/util/src/platform.c:219]: Oops, I meant hse_cpu_init() and timer_init()...   We're not not using timers in vlb_init() right now, but I don't wish to preclude it...

File: tests/unit/mpool/mblock_test.c (hunk 4 above)
Comment (beckerg) [reply tests/unit/mpool/mblock_test.c:351]: Ah, that old irritant.  The sooner we get rid of all our custom integer type names the better IMO.

**Comment by beckerg** [comment]
Won't that spew a bunch of error message to the cli user rather than just the one that bubbled up?

**Comment by tristan957** [comment]
Alex and I discussed a solution which I will push up here after implementing.

**Review by alexttx — APPROVED** [review]
